script/constructron.lua

```diff
@@ -799,46 +734,79 @@ me.conditions = {
     ---@param position MapPosition
     ---@return boolean
     position_done = function(job, position) -- this is condition for action "go_to_position"
-        local constructrons = job.constructrons
-        debug_lib.VisualDebugText("Moving to position", constructrons[1], -3, 1)
-        if constructrons[1].valid then
-            if (chunk_util.distance_between(constructrons[1].position, position) > 5) then -- the condition
-                if not ((job.attempt or 0) >= 3) then
-                    return false  -- condition is not met
-                else
-                    return "graceful_wrapup" -- unable to reach destination
-                end
+        local constructron = job.constructrons[1]
+        debug_lib.VisualDebugText("Moving to position", constructron, -3, 1)
+
+        local ticks = (game.tick - job.start_tick)
+        if ticks > 119 then return false end -- not enough time (two seconds) since last check
+
+        local distance_from_pos = chunk_util.distance_between(constructron.position, position)
+        if (distance_from_pos < 5) then return true end -- condition is satisfied
+
+        if not job.path_active then -- check if the path is active
+            if job.request_pathid and not global.pathfinder_requests[job.request_pathid] then -- check that there is a request
+                me.actions[job.action](job, table.unpack(job.action_args or {})) -- there is no request, request a path.
+            end
+            debug_lib.VisualDebugText("Path not active", constructron, 0.4, 3)
+            return false -- condition is not met
+        end
+
+        if job.landfill_job then -- is this a landfill job?
+            if not constructron.logistic_cell.logistic_network.can_satisfy_request("landfill", 1) then
+                -- !! Logic gap - Constructrons will return home even if there is other entities to build.
+                me.graceful_wrapup(job) -- no landfill left.. leave
+                return false
             end
-        else
-            return true  -- leader is invalidated.. skip action
         end
-        me.enable_roboports(constructrons[1].grid)
-        return true -- condition is satisfied
+
+        if job.attempt > 3 then
+            me.graceful_wrapup(job)
+            return false
+        end
+
+        if not constructron.autopilot_destination then -- path lost recovery
+            job.start_tick = game.tick
+            me.actions[job.action](job, table.unpack(job.action_args or {}))
+            return false
+        end
+
+        local mvmt_last_distance = job.mvmt_last_distance
+        local distance = chunk_util.distance_between(constructron.position, constructron.autopilot_destination)
+        if mvmt_last_distance and (mvmt_last_distance - distance < 2) and constructron.speed < 0.1 then -- check that movement has progressed at least two tiles in the last 10 seconds.

```

Comment:
The comment on this condition is the wrong way around. You check if the ctron has moved less than 2 tiles -> rerequest a path